connector/signaltometrics/connector.go

```diff
@@ -0,0 +1,281 @@
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/elastic/opentelemetry-collector-components/connector/signaltometricsconnector/internal/aggregator"
+	"github.com/elastic/opentelemetry-collector-components/connector/signaltometricsconnector/internal/model"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/ottldatapoint"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/ottllog"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/ottlspan"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/sampling"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+)
+
+const (
+	ephemeralResourceKey = "signaltometrics_ephemeral_id"
+)
+
+type signalToMetrics struct {
+	component.StartFunc
+	component.ShutdownFunc
+
+	next           consumer.Metrics
+	spanMetricDefs []model.MetricDef[ottlspan.TransformContext]
+	dpMetricDefs   []model.MetricDef[ottldatapoint.TransformContext]
+	logMetricDefs  []model.MetricDef[ottllog.TransformContext]
+
+	// ephemeralID will be removed, see: https://github.com/elastic/opentelemetry-collector-components/issues/159
+	ephemeralID string
+}
+
+func (sm *signalToMetrics) Capabilities() consumer.Capabilities {
+	return consumer.Capabilities{MutatesData: false}
+}
+
+func (sm *signalToMetrics) ConsumeTraces(ctx context.Context, td ptrace.Traces) error {
+	var multiError error
+	processedMetrics := pmetric.NewMetrics()
+	processedMetrics.ResourceMetrics().EnsureCapacity(td.ResourceSpans().Len())
+	aggregator := aggregator.NewAggregator[ottlspan.TransformContext](processedMetrics)
+
+	for i := 0; i < td.ResourceSpans().Len(); i++ {
+		resourceSpan := td.ResourceSpans().At(i)
+		resourceAttrs := resourceSpan.Resource().Attributes()
+		for j := 0; j < resourceSpan.ScopeSpans().Len(); j++ {
+			scopeSpan := resourceSpan.ScopeSpans().At(j)
+			for k := 0; k < scopeSpan.Spans().Len(); k++ {
+				span := scopeSpan.Spans().At(k)
+				spanAttrs := span.Attributes()
+				adjustedCount := calculateAdjustedCount(span.TraceState().AsRaw())
+				for _, md := range sm.spanMetricDefs {
+					filteredSpanAttrs := getFilteredAttributes(spanAttrs, md.Attributes)
+					if filteredSpanAttrs.Len() != len(md.Attributes) {
+						// If all the configured attributes are not present in
+						// source metric then don't count them.
+						continue
+					}
+					var filteredResAttrs pcommon.Map
+					if len(md.IncludeResourceAttributes) > 0 {
+						filteredResAttrs = getFilteredAttributes(resourceAttrs, md.IncludeResourceAttributes)
+					} else {
+						// Copy resource attrs to avoid mutating data
+						filteredResAttrs = pcommon.NewMap()
+						resourceAttrs.CopyTo(filteredResAttrs)
+					}
```

Comment:
I think we should find a way to do this once per ResourceSpans per metric definition

Reply:
In a follow-up, I will be adding collector's service information as observer (for single-writer). This will be added based on metric definition meaning we will have to allocate the underlying array accordingly for as per configuration in metric definition. We can keep a copy of the resource attributes and avoid the necessary copy, but I am not sure caching filtered attributes at the current point would give us any benefits.  Does it sound good to revisit this on the PR for the single-writer?